memif-attach/main.go

```diff
@@ -18,61 +18,120 @@ package main
 
 import (
 	"context"
+	"net"
 	"time"
-
+	// "io"
+	"encoding/binary"
+	"bufio"
+	"bytes"
+	"unsafe"
 	"git.fd.io/govpp.git/api"
 	"github.com/edwarnicke/govpp/binapi/memif"
+	interfaces "github.com/edwarnicke/govpp/binapi/interface"
+	"github.com/edwarnicke/govpp/binapi/interface_types"
 	"github.com/edwarnicke/log"
 	"github.com/edwarnicke/vpphelper"
 )
 
+type MemifMsg struct {
+	HelloMsgType HelloMsg
+	InitMsgType InitMsg
+}
+// HelloReplyMsg type
+type HelloMsg struct {
+	Max_region uint16 
+	Max_m2s_ring uint16 
+	Max_s2m_ring uint16
+	Min_version uint16 
+	Max_version uint16
+	Max_log2_ring_size uint8 
+	Name [32]uint8
+}
+
+// InitMsg
+const MEMIF_SECRET_SIZE = 24
+type MemifInterfaceMode int
+const (
+	MEMIF_INTERFACE_MODE_ETHERNET MemifInterfaceMode = iota
+	MEMIF_INTERFACE_MODE_IP
+	MEMIF_INTERFACE_MODE_PUNT_INJECT
+)
+type InitMsg struct {
+	Version uint16
+	Id uint32
+	Mode MemifInterfaceMode
+	Secret  [MEMIF_SECRET_SIZE]uint8
+	Name [32]uint8
+}
+// UnmarshalBinary Function
+func (replyMsg *HelloMsg) UnmarshalBinary(data []byte) error {
+	buf := bytes.NewReader(data)
+	err := binary.Read(buf, binary.BigEndian, replyMsg)
+	return err
+}
 func main() {
 	ctx, cancel := context.WithCancel(context.Background())
-
+	
 	// Connect to VPP with a 1 second timeout
 	connectCtx, _ := context.WithTimeout(ctx, time.Second)
 	conn, vppErrCh := vpphelper.StartAndDialContext(connectCtx)
 	exitOnErrCh(ctx, cancel, vppErrCh)
-
+	
 	// Create a RPC client for the memif api
 	isClient := true
 	// Add a memif socket
-	socketID, _ := createMemifSocket(ctx, conn)
+	socketID, socketAddr, _:= createMemifSocket(ctx, conn)
 	// Create a memif interface
-	mode := memif.MEMIF_MODE_API_IP
-	createMemif(ctx, conn, socketID, mode, isClient)
-	// Delete a memif socket
+	createMemif(ctx, conn, socketID, isClient)
+	
+	// Dump memif info
 	dumpMemif(ctx, conn)
+	
+	// connect to VPP
+	conn_client, err := net.Dial("unixpacket", socketAddr)
+	if err != nil {
+		log.Entry(ctx).Fatalln("ERROR: connect to VPP master failed:", err)
+	}
+	// read hello message from master vpp
+	helloMsg := &HelloMsg{}
+	reader := bufio.NewReader(conn_client)
+	helloErr := binary.Read(reader, binary.BigEndian, helloMsg)
+	if helloErr != nil {
+		log.Entry(ctx).Fatalln("ERROR: read from VPP master hello message failed:", err)
+	}
+	log.Entry(ctx).Infof("max_region: %v\n"+"size of hellomsg: %v\n",
```

Comment:
And I tried to log the value of max_region and size of hello message, and the size of the hello message do not seem to make sense to me.

Reply:
Ah, I forgot to dereference the hellomsg pointer. Now I get 44 bytes size for hellomsg which should make some sense now.